File: solutions/python/bob/1/bob.py


def response( hey_bob):

    hey_bob = str(hey_bob).strip()

    if hey_bob == "":
        return "Fine. Be that way!"

    if hey_bob.isupper() and hey_bob[-1] == '?':
        return "Calm down, I know what I'm doing!"

    if hey_bob[-1] == '?':
        return "Sure."

    if hey_bob.isupper():
        return "Whoa, chill out!"

    return "Whatever."

File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where
import Data.Char (isAsciiLower, isAsciiUpper, isSpace)

responseFor :: String -> String
responseFor xs
  | is_yelling && is_question = "Calm down, I know what I'm doing!"
  | is_yelling = "Whoa, chill out!"
  | is_question = "Sure."
  | null no_space = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    is_yelling = not (null upper_only) && null lower_only 
    is_question = not (null no_space) && last no_space == '?'

    no_space = filter (\x -> not (isSpace x)) xs
    upper_only = filter isAsciiUpper no_space
    lower_only = filter isAsciiLower no_space
